api/http-api.md

```diff
@@ -0,0 +1,176 @@
+instead of as an HTTP header. This makes URL's large and a bit unwieldy, but calls like from `curl` would have to have
+`-H Temporal-Namespace: <my-ns>` which is even more unwieldy.
+
+The bodies of requests are proto JSON of the request messages and responses are full proto JSON of the response
+messages. While we try to somewhat follow REST principles, this is not a full RESTful API necessarily. Developers adding
+new API will need to add the gRPC gateway annotations for their RPC calls.
+
+For the initial version, we will only be doing workflow service. Operator service and cloud APIs may come soon after
+though.
+
+### Rules
+
+* Naming
+  * Multi-word URL paths use kebab-case
+  * Multi-word query parameters use camelCase (only first letter of abbreviations capitalized if not first word)
+  * Do not assume any names reserved. So we can't have `/workflows/{workflow_id}` and `/workflows/general-call` in case
+    there's a workflow named `general-call`.
+  * Use plural nouns where an identifier may be provided to differentiate
+    * But singular nouns are not _required_ in other places, can still be plural
+* Response codes
+  * 201 success if a creation
+  * 202 if action eventually consistent
+  * 200 other all other successes
+  * Do not use 204 for no response, even if there's no response because we may add response later, just 200 everything
+* Methods:
+  * `GET` if read-only
+    * Noun as last URL term
+    * Use query parameters for commonly set items to avoid relying on request bodies (but still rely on bodies as
+      needed). This may require some gRPC gateway customization.
+  * `PUT` if strongly consistent add/update on fixed-URL w/ idempotency
+    * Noun as last URL term
+  * `DELETE` if item is actually deleted (i.e. unreachable from `GET`)
+    * Noun as last URL term
+  * `POST` for everything else
+    * Noun as last URL term if for a create, verb if for an action
+  * `PATCH` is not needed because we don't adhere to common CRUD
+    * TODO(cretz): Needs discussion
+
+### RPCs
+
+Below is list of gRPC calls to HTTP calls so far and some notes/questions beneath each. Some notes/questions beneath one
+actually apply to many but we are not repeating ourselves too much here.
+
+* Namespaces:
+  * `RegisterNamespace` - `POST /api/v1/namespaces`
+    * 202 on success since eventually consistent
+  * `DescribeNamespace` - `GET /api/v1/namespaces/{namespace}`
+  * `ListNamespaces` - `GET /api/v1/namespaces`
+  * `UpdateNamespace` - `PUT /api/v1/namespaces/{namespace}`
+    * TODO(cretz): Confirm strongly consistent and idempotent or not
+  * `DeprecateNamespace` - `DELETE /api/v1/namespaces/{namespace}`
+    * Is this eventually consistent? Will determine response code
+    * Does this prevent returning from `GET`? Will determine HTTP method
+* Workflows:
+  * `StartWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}`
+    * 204 on success
```

Comment:
You said no 204, is this a mistake? I think you meant 201.

Reply:
Yes, typo! Will fix.

Reply:
Fixed